### PR TITLE
Add networking support

### PR DESCRIPTION
**Pull request by raharper**
Implement the network connections, networks and nics configuration.

Network supports three types: user, bridge and mcast

'user'   - qemu slirp networking
'bridge' - host bridge and tun/tap interfaces (requires sudo privs)
'mcast'  - qemu socket networking using multicast routing between VMs
           using the same multicast network and ports (no sudo privs
           required)

### CODE REVIEW

File: run.go [hunk 1]
```diff
@@ -106,6 +106,37 @@ func doRun(ctx *cli.Context) error {
...
+					if network.Name == networkName {
+						found = true
+						break
+					}
+				}
+				if !found {
+					return fmt.Errorf("Connection specified unknown network: %s", networkName)
+				}
+				for nidx := range vm.Nics {
+					found = false
```
Comment (hallyn) [inline run.go]: This doesn't look right.  Did you mean the 'found = false' to happen outside of the for nidx := range vm.Nics loop?  Or am I misunderstanding the flow?

It looks like if vm.Nics is empty, you *should* return an error for the connection nic being undefined, but in fact 'found' will be true at line 134 and you'll proceed.

File: run.go [hunk 2]
```diff
@@ -114,11 +145,28 @@ func doRun(ctx *cli.Context) error {
...
+			vmNic := vm.Nics[nidx]
+			log.Debugf("Adding nic %s network %s", vmNic.ID, vmNic.Network)
+			var nicNetwork NetworkDef
+			for netidx := range suite.Networks {
+				network := suite.Networks[netidx]
+				if network.Name == vmNic.Network {
+					nicNetwork = network
+					break
+				}
+			}
```
Comment (hallyn) [inline run.go:159]: Can we assert here that nicNetwork was in fact found?

File: vmnics.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	return string(b)
+}
+
+const LinuxIfnameLength int = 15
+
+// generate a random Linux network interface name.
+// Allow specifying a prefix
+func randomNicIFName(prefix string) string {
+	prefixLength := len(prefix)
+	if prefixLength >= LinuxIfnameLength {
```
Comment (hallyn) [inline vmnics.go]: Shall we go a step further and require at least space for two random characters?  The pain of debugging subtle breakage when ifnames become likely to be the same is probably going to be worse than that of anyone wanting to use 14 character prefix.

File: vmnics.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+		return randomString(LinuxIfnameLength)
+	}
+	randLength := LinuxIfnameLength - prefixLength
+	return prefix + randomString(randLength)
+}
+
+// The VMNic lifecycle: first you create one with vm.AddNic().
+// The you call Setup on it to create the tap.  Then you call
```
Comment (hallyn) [inline vmnics.go]: *Then*

File: vmnics.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+		NetAddr:    netdef.Address,
+		BootIndex:  nicdef.BootIndex,
+		Ports:      nicdef.Ports,
+	}
+	log.Debugf("AddNic: Nic:%s Slot:%s -> IFName:%s Type:%s", nic.ID, nic.BusAddr, nic.IFName, nic.NetType)
+	v.opts.NICs = append(v.opts.NICs, &nic)
+	return nil
+}
+
+var macOffset int = 0
```
Comment (hallyn) [inline vmnics.go:75]: Just a note that this works well so long as we continue to start clusters as a whole, but won't work if we start having separate programs start the nics.  (Unless we use an API to request a single daemon set up the hwaddrs)

File: run.go (hunk 2 above)
Comment (raharper) [reply run.go:159]: Yes, I'll check it

File: vmnics.go (hunk 3 above)
Comment (raharper) [reply vmnics.go]: Sure

File: vmnics.go (hunk 4 above)
Comment (raharper) [reply vmnics.go]: +1

File: vmnics.go (hunk 5 above)
Comment (raharper) [reply vmnics.go:75]: Yes, I think we should:

1) switch to starting clusters-as-a-whole via API
2) enable API to start/stop VMs within a cluster

File: run.go (hunk 1 above)
Comment (raharper) [reply run.go]: I'll review.  I didn't test with a nic-less VMs, so that's likely a bug I'll fix.

File: vmnics.go (hunk 5 above)
Comment (hallyn) [reply vmnics.go:75]: Sounds good.

File: run.go (hunk 1 above)
Comment (raharper) [reply run.go]: Thanks, found the issue.

**Pull request merged**